### PR TITLE
[BEAM-9444] Use GCP BOM to set package versions

### PR DESCRIPTION
**Pull request by kileys**
Use GCP libraries bom to set versions for GCP related libraries

Updated version of #11156
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by kileys** [comment]
Run Java examples on Dataflow Java 11

**Comment by kileys** [comment]
Run Dataflow ValidatesRunner Java 11

**Comment by kileys** [comment]
Run Dataflow ValidatesRunner

**Comment by kileys** [comment]
Run Java PostCommit

**Comment by kennknowles** [comment]
Run Spotless PreCommit

**Comment by kennknowles** [comment]
I don't see any reason for that strange error in spotless.

**Comment by kileys** [comment]
Run Python PreCommit

**Comment by kileys** [comment]
Run Spark ValidatesRunner

**Comment by kileys** [comment]
Run Dataflow ValidatesRunner

**Comment by kileys** [comment]
@suztomo are there other tests I should run to verify this? Linkage checker runs without any new errors

**Comment by kileys** [comment]
R: @kennknowles, @suztomo, @lukecwik

**Comment by kileys** [comment]
The Dataflow validates runner tests are also failing on master

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -21,6 +21,8 @@ package org.apache.beam.gradle
 import com.github.jengelman.gradle.plugins.shadow.tasks.ShadowJar
 import groovy.json.JsonOutput
 import groovy.json.JsonSlurper
+import java.util.concurrent.atomic.AtomicInteger
```
Comment (suztomo) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:24]: I didn't expect to use AtomicInteger. Do you really need this?

**Review by suztomo — APPROVED** [review]
> are there other tests I should run to verify this?

Not that I know of.

**Comment by lukecwik** [comment]
Also, it seems as though existing modules that don't add a platform level dep to the GCP libraries bom will get some version of the dep since we have cleared the version.

File: examples/java/build.gradle [hunk 2]
```diff
@@ -48,6 +48,7 @@ configurations.sparkRunnerPreCommit {
 }
 
 dependencies {
+  compile platform(library.java.google_cloud_platform_libraries_bom)
```
[the remark this answers is not in the document]
Comment (kileys) [reply examples/java/build.gradle]: Updated to enforcedPlatform and going to update the rest of the dependencies in a follow up

**Review by kennknowles — APPROVED** [review]
LGTM

**Comment by kileys** [comment]
Run Java PreCommit

**Comment by kileys** [comment]
Run Python PreCommit

**Comment by kileys** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by kennknowles** [comment]
It does seem that if a module had any of these deps but forgot the `enforcedPlatform` line then it would go back to getting the arbitrary latest version. I understand fixing that is in a follow-up.

Here's what I did to view the dep changes:

```
# gcp is GNU cp here
mkdir ~/tmp/deps-before
git checkout github/master
./gradlew dependencyReport
find . -name dependencies.txt -exec gcp --parents \{\} ~/tmp/deps-before/ \;

mkdir ~/tmp/deps-after
git checkout github/pr/13075
./gradlew dependencyReport
find . -name dependencies.txt -exec gcp --parents \{\} ~/tmp/deps-after/ \;

diff --brief --recursive ~/tmp/deps-before ~/tmp/deps-after

diff --unified=0 --recursive ~/tmp/deps-before ~/tmp/deps-after
```

**Review by kennknowles — APPROVED** [review]
I do not have the capacity to manually determine the possible consequences of the dep changes, which do affect modules not listed (for example autovalue annotations are downgraded a couple patch versions).

I was checking just for modules that were accidentally "upgraded" to the latest by not having a version or that were accidentally downgraded because the BOM reduced versions they depended on that had nothing to do with GCP. I did not find any problems, but it is too much content to be certain. Still, it is surely an improvement.

**Pull request merged**